### ISSUE TITLE
Input of the classifier in the last layer is wrong

### ISSUE THREAD

**Issue by enginin22** [tintn/vision-transformer-from-scratch#1]
# Improvement
In class `ViTForClassfication` from `vit.py`, replace  `logits = self.classifier(encoder_output[:, 0])` with  `logits = self.classifier(encoder_output[:, 0, :]) `

# Reason
The shape of `encoder_output`should be `[batch_size, number_of_patch+1, embedding_size]` and have 3 dimensions. When using **CLS**,  the first token of `encoder_output` , it is `encoder_output[:, 0, :]`

**Comment by tintn** [comment]
Hi @veraz00, thanks for raising the issue. 

The two ways are actually equivalent because PyTorch implicitly selects all elements from the later axes. To verify it, you can run the following:
```python
a = torch.rand(10, 20, 30)
print((a[:, 0] == a[:, 0, :]).all())
# Output:
# tensor(True)
```

However, It's true that I better keep the axes explicitly specified. I will make the change soon.
